### ISSUE TITLE
Can't set time with get_output_stream

### ISSUE THREAD

**Issue by u3shit** [rubyzip/rubyzip#503]
I'm trying to put an entry with some arbitrary timestamp into the zip file, but it doesn't work, it always writes `Time.now`:
```ruby
require 'zip'
time = Time.new '2020-01-01' # some random date
Zip::File.open 'foo.zip', create: true do |zip|
  e = Zip::Entry.new zip.name, 'some_file'
  e.time = Zip::DOSTime.from_time time
  zip.get_output_stream(e) {|s| s.puts 'foobar' }
end
```
But unzip reports the current date:
```
$ unzip -l foo.zip 
Archive:  foo.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
        7  11-06-2021 17:56   some_file
---------                     -------
        7                     1 file
```

It looks like `File#get_output_stream` adds a `StreamableStream` to `@entry_set`, but later in `OutputStream#put_next_entry` it checks `entry_name.kind_of?(Entry)`, which will be false and it creates a new `Entry` ignoring everything I've set up in the entry. The same happens if I don't create the entry manually but pass the time to `#get_output_stream`.
`File#add` works correctly, but it requires creating a temporary file and manually setting the last modification date on it.
Using ruby 3.0.2p107 and rubyzip 2.3.2.

**Comment by hainesr** [comment]
Hi @u3shit, thanks for reporting this.

I think I see what's going wrong here: presumably the type mismatch between `StreamableStream` and `Entry`. I wasn't around when this was written, so I don't know the subtleties of why `StreamableStream` is a delegate, rather than a subclass, of `Entry`, but just naively widening the `kind_of?` check you mention above causes plenty of other errors.

I can see a way to a fix though, so I'll get this sorted as soon as I can.

**Comment by hainesr** [comment]
Hi @u3shit, I've implemented a fix for this in #504. It fixes the example you posted above (for me), so if you have time to test as well that would be useful, thanks.

**Comment by u3shit** [comment]
Thanks, it looks like it's working with my code (after I fixed the 3.0 api break with optional parameters, uh-oh. But at least it looks like I don't even have to construct an `Entry` manually now since I can pass time to `get_output_stream` without having to type countless `nil`s)

**Comment by hainesr** [comment]
Thanks for checking! ⚡ 

I am a bit nervous about the API changes, but hopefully people will see the benefits of the more modern style. As you say - not so many `nil`s...
